src/main/java/net/rubygrapefruit/platform/internal/jni/AbstractFileEventFunctions.java

```diff
@@ -46,25 +50,47 @@ public AbstractWatcherBuilder(FileWatcherCallback callback) {
      * Configures a new watcher using a builder. Call {@link AbstractWatcherBuilder#start()} to
      * actually start the {@link FileWatcher}.
      */
-    public abstract AbstractWatcherBuilder newWatcher(FileWatcherCallback callback);
+    public abstract AbstractWatcherBuilder newWatcher(BlockingQueue<FileWatchEvent> queue);
 
     protected static class NativeFileWatcherCallback {
-        private final FileWatcherCallback delegate;
 
-        public NativeFileWatcherCallback(FileWatcherCallback delegate) {
-            this.delegate = delegate;
+        private final BlockingQueue<FileWatchEvent> eventQueue;
+
+        public NativeFileWatcherCallback(BlockingQueue<FileWatchEvent> eventQueue) {
+            this.eventQueue = eventQueue;
         }
 
         // Called from the native side
         @SuppressWarnings("unused")
-        public void pathChanged(int type, String path) {
-            delegate.pathChanged(FileWatcherCallback.Type.values()[type], path);
+        public void pathChanged(int typeIndex, String path) throws InterruptedException {
+            FileWatchEvent.Type type = FileWatchEvent.Type.values()[typeIndex];
+            if (type == FileWatchEvent.Type.OVERFLOWED) {
+                signalOverflow(path);
+            } else {
+                queueEvent(new ChangeEvent(type, path), false);
+            }
         }
 
         // Called from the native side
         @SuppressWarnings("unused")
-        public void reportError(Throwable ex) {
-            delegate.reportError(ex);
+        public void reportError(Throwable ex) throws InterruptedException {
+            queueEvent(new ErrorEvent(ex), true);
```

Comment:
Shouldn't we clear the queue on error as well? It doesn't really make sense to process any other events before the error, right?

Reply:
I'm not sure. We clear the events on overflow because with an overflow we already know we lost some events, so it's just a matter of quantity. There is no such necessity when errors arise.

If we cleared the queue on error, we'd assume that errors invalidate all previous events. While that's the approach we'll take with Gradle, it's not necessarily the only possible way to handle this.

Also, if we'd clear the queue on an error, should we also clear on an unknown event?

I think it's better as it is.